### DISCUSSION TITLE
Cannot register menu items asynchronously

### DISCUSSION THREAD

**Discussion by bendavis78** [eclipse-theia/theia#8815] [Q&A]
When creating a MenuContribution + CommandContribution, I am unable to use asynchronous logic inside the `registerCommands` and `registerMenus` functions. When you use async in either function, the menus are not registered properly (they simply don't appear).

You can reproduce this by starting from generator-theia-extension using the "widget" template. Update the `reigsterMenus` or `registerCommands` functions to register asynchronously, like so:

```typescript
    registerCommands(commands: CommandRegistry): void {
        setTimeout(() => {
            commands.registerCommand(TheiaTestCommand, {
                execute: () => super.openView({ activate: false, reveal: true })
            });
        });
    }

    registerMenus(menus: MenuModelRegistry): void {
        setTimeout(() => {
            super.registerMenus(menus);
        });
    }
```

To answer the question of "why would you want to do this", I need to create menu items dynamically based on a server-side configuration. That configuration is fetched asynchronously via JSONRPC.

**Comment by kittaakos** [comment]
> menu items dynamically based on a server-side configuration

See here: https://community.theia-ide.org/t/menuactions-based-on-data-recieved-from-backend/1200/2

**Reply by bendavis78** [reply]
@kittaakos I tried that within the `registerMenus` method, but it does not work within an async context. 

```typescript
    registerMenus(menus: MenuModelRegistry): void {
        setTimeout(() => {
            this.toDisposeBeforeMenuUpdate.dispose();
            this.toDisposeBeforeMenuUpdate.push(this.commandRegistry.registerCommand({ id: 'foo-id' }, { execute: () => { console.log('command executed') }}));
            this.menuModel.registerMenuAction(CommonMenus.HELP, { commandId: 'foo-id', label: 'Foo Command' });
            this.toDisposeBeforeMenuUpdate.push(Disposable.create(() => this.menuModel.unregisterMenuNode('foo-id')));
        });
    }
```

**Reply by bendavis78** [reply]
~~Actually, never mind -- I put it in a @postConstruct method and it worked!~~

**Reply by bendavis78** [reply]
@kittaakos Scratch that, it didn't work when I increased the timeout delay. Seems to be a race condition somewhere:

```typescript
    @postConstruct()
    protected init(): void {
        setTimeout(() => {
            this.commandRegistry.registerCommand({ id: 'foo-id' }, { execute: () => { console.log('command executed') }});
            this.menuModel.registerMenuAction(CommonMenus.HELP, { commandId: 'foo-id', label: 'Foo Command' });
        }, 1000);
    }
```

The command is registered successfully, however the menu item does not appear. Is there a render function somewhere that's not being called?

**Comment by bendavis78** [comment]
I did some digging, and the issue appears to be with [core/lib/browser/menu/browser-menu-plugin:292](https://github.com/eclipse-theia/theia/blob/f1e37fbdbe9cf0bdcd30f58d2c8fbf6d6808419e/packages/core/src/browser/menu/browser-menu-plugin.ts#L292):
```typescript
                if (!(commands.isVisible(node.action.commandId) && (!when || contextKeyService.match(when)))) {
```
The `commands.isVisible(node.action.commandId)` returns `false` because the command is not registered with the menu. The only place I can see where commands are registered with the menu is in `MenuCommandRegistry.registerActionMenu()`, and this is only called during `BrowserMenuContribution.onStart()`.

I feel this is a bug in theia. I've reported the issue here: https://github.com/eclipse-theia/theia/issues/8817
